### PR TITLE
WDPD-33 Credit card configuration

### PR DESCRIPTION
**Pull request by ps-ohanel**
### This PR

* adds Credit Card payment method configuration
* shows Credit Card payment method config information on the checkout page

### Note

* CR comments regarding this code are implemented

### How to Test

* Login to the admin and activate the plugin.
* Go to Payment methods and select "Wirecard Credit Card".
* You are able to setup all needed fields for Credit card
* As a customer: You can see in checkout process Credit Card payment method. When selecting, debug info of payment method fields is displayed

### Jira Links
* https://jira.parkside.at/browse/WDPD-33

### CODE REVIEW

File: Core/OxidEE_Events.php [hunk 1]
```diff
@@ -128,6 +127,26 @@ private static function _extendPaymentMethodTable()
...
+        $sQueryAddMaxLimit = "ALTER TABLE " . self::PAYMENT_TABLE .
+            " ADD COLUMN `WDOXIDEE_NON_THREE_D_MAX_LIMIT` double NOT NULL";
+        self::_addColumnIfNotExists(self::PAYMENT_TABLE, 'WDOXIDEE_NON_THREE_D_MAX_LIMIT', $sQueryAddMaxLimit);
+
+        $sQueryAddMinLimit = "ALTER TABLE " . self::PAYMENT_TABLE .
+            " ADD COLUMN `WDOXIDEE_THREE_D_MIN_LIMIT` double NOT NULL";
+        self::_addColumnIfNotExists(self::PAYMENT_TABLE, 'WDOXIDEE_THREE_D_MIN_LIMIT', $sQueryAddMinLimit);
+
+        $sQueryDefCurrency = "ALTER TABLE " . self::PAYMENT_TABLE . "
+            ADD COLUMN `WDOXIDEE_DEFAULT_CURRENCY` varchar(128) default '' NOT NULL";
```
Comment (tomazpu) [inline Core/OxidEE_Events.php]: I suggest naming it limits currency

File: out/blocks/wd_admin_payment_main_form.tpl [hunk 2]
```diff
@@ -119,6 +119,57 @@
         [{oxinputhelp ident="config_three_d_merchant_secret_desc"}]
     </td>
   </tr>
+
```
Comment (tomazpu) [inline out/blocks/wd_admin_payment_main_form.tpl]: Can u extend the base template and create a new creditcard template

File: Extend/Payment_Main.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+     * Executes parent::render(), creates a currency array
+     *
+     * @return string
+     *
+     */
+    public function render()
+    {
+        $sTemplate = parent::render();
+        $oConfig = Registry::getConfig();
+        if ($this->_aViewData["oxid"] == "wdcreditcard") {
```
Comment (tomazpu) [inline Extend/Payment_Main.php]: I suggest getting the payment method name, then having a getter in each payment method where you can get the "additional fields" and in the abstract payment_method you just return an empty array. Then here all you do is merge _aViewData with the getter function

Comment (ps-ohanel) [reply Extend/Payment_Main.php]: This is addressed in Pull Request #32 . Let's prioritise that one and keep this open until #32 is merged.

File: out/blocks/wd_admin_payment_main_form.tpl (hunk 2 above)
Comment (ps-ohanel) [reply out/blocks/wd_admin_payment_main_form.tpl]: This is addressed in Pull Request #32 . Let's prioritise that one and keep this open until #32 is merged.

File: Core/OxidEE_Events.php (hunk 1 above)
Comment (ps-ohanel) [reply Core/OxidEE_Events.php]: renamed

File: Model/Credit_Card_Payment_Method.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                'title'       => Helper::translate('config_ssl_max_limit'),
+                'description' => Helper::translate('config_ssl_max_limit_desc'),
+            ],
+            'threeDMinLimit' => [
+                'type'        => 'text',
+                'field'       => 'oxpayments__wdoxidee_three_d_min_limit',
+                'title'       => Helper::translate('config_three_d_min_limit'),
+                'description' => Helper::translate('config_three_d_min_limit_desc'),
+            ],
+            'basket' => [
```
Comment (ps-cgrach) [inline Model/Credit_Card_Payment_Method.php]: I think basket is the wrong key for the currency

File: default_payment_config.xml [hunk 5]
```diff
@@ -17,4 +17,26 @@
...
         <wdoxidee_additional_info>1</wdoxidee_additional_info>
     </payment>
+    <payment>
+        <oxid>wdcreditcard</oxid>
+        <oxactive>0</oxactive>
+        <oxtoamount>1000000</oxtoamount>
+        <oxdesc>Wirecard Kreditkarte</oxdesc>
+        <oxdesc_1>Wirecard Credit Card</oxdesc_1>
+        <wdoxidee_logo>creditcard.png</wdoxidee_logo>
+        <wdoxidee_transactionaction>purchase</wdoxidee_transactionaction>
```
Comment (ps-cgrach) [inline default_payment_config.xml]: didn't we change that to `pay` and `reserve` already?

Comment (ps-ohanel) [reply default_payment_config.xml]: changed to pay after rebase

File: Model/Credit_Card_Payment_Method.php (hunk 4 above)
Comment (ps-ohanel) [reply Model/Credit_Card_Payment_Method.php]: fixed key naming

File: Model/Credit_Card_Payment_Method.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        $oConfig = new Config(
+            $oPayment->oxpayments__wdoxidee_apiurl->value,
+            $oPayment->oxpayments__wdoxidee_httpuser->value,
+            $oPayment->oxpayments__wdoxidee_httppass->value
+        );
+        $oPaymentMethodConfig = new PaymentMethodConfig(
+            self::getName(),
+            $oPayment->oxpayments__wdoxidee_maid->value,
+            $oPayment->oxpayments__wdoxidee_secret->value
+        );
```
Comment (tomazpu) [inline Model/Credit_Card_Payment_Method.php]: Im missing the the setting of the 3dmaid and 3dsecret

Comment (tomazpu) [reply Model/Credit_Card_Payment_Method.php]: Also for credit card you need to use CreditCardConfig

Comment (ps-ohanel) [reply Model/Credit_Card_Payment_Method.php]: added 3dmaid and 3dsecret and switched to CreditCardConfig.

**Pull request merged**